Quadriga/src/main/java/edu/asu/spring/quadriga/web/workspace/AddTextController.java

```diff
@@ -0,0 +1,139 @@
+    @RequestMapping(value = "/auth/workbench/workspace/{projectid}/{workspaceid}/addtext", method = RequestMethod.GET)
+    public ModelAndView addTextFileForm(@PathVariable("workspaceid") String workspaceid,
+            @PathVariable("projectid") String projid) throws QuadrigaStorageException, QuadrigaAccessException {
+
+        ModelAndView model = new ModelAndView("auth/workbench/workspace/addtext");
+        model.getModelMap().put("textfile", textFileFactory.createTextFileObject());
+        model.getModelMap().put("workspaceId", workspaceid);
+        model.getModelMap().put("myProjectId", projid);
+        model.getModelMap().put("success", "0");
+        return model;
+    }
+
+    /**
+     * @param resp
+     *            Generic HTTP Response Object
+     * @param txtFile
+     *            Web backing domain object to store the Text Details
+     * @param result
+     *            Generic BingingResults object to hold validation errors
+     * @param workspaceid
+     *            Path Variable to retrieve the associated Workspace id
+     * @return
+     *          Returns the workspace details page if success else returns the add text page.
+     * @throws QuadrigaStorageException
+     * @throws QuadrigaAccessException
+     * @throws IOException
+     */
+    @AccessPolicies({
+            @ElementAccessPolicy(type = CheckedElementType.WORKSPACE, paramIndex = 4, userRole = {
+                    RoleNames.ROLE_WORKSPACE_COLLABORATOR_ADMIN, RoleNames.ROLE_WORKSPACE_COLLABORATOR_EDITOR }),
+            @ElementAccessPolicy(type = CheckedElementType.PROJECT, paramIndex = 5, userRole = {
+                    RoleNames.ROLE_COLLABORATOR_ADMIN, RoleNames.ROLE_PROJ_COLLABORATOR_ADMIN,
+                    RoleNames.ROLE_PROJ_COLLABORATOR_CONTRIBUTOR }) })
+    @RequestMapping(value = "/auth/workbench/workspace/{projectid}/{workspaceid}/addtext", method = RequestMethod.POST)
+    public ModelAndView saveTextFileForm(HttpServletResponse resp,
+            @Validated @ModelAttribute("textfile") TextFile txtFile, BindingResult result,
+            @PathVariable("workspaceid") String workspaceid, @PathVariable("projectid") String projid,
+            RedirectAttributes redirectAttributes)
+                    throws QuadrigaStorageException, QuadrigaAccessException, IOException {
+
+        ModelAndView model;
+        if (result.hasErrors()) {
+            model = new ModelAndView("auth/workbench/workspace/addtext");
+            model.getModelMap().put("textfile", txtFile);
+            model.getModelMap().put("workspaceId", workspaceid);
+            model.getModelMap().put("myProjectId", projid);
+            model.getModelMap().put("success", "0");
+        } else {
+            model = new ModelAndView("redirect:/auth/workbench/workspace/workspacedetails/" + workspaceid);
+
+            model.getModelMap().put("workspaceId", workspaceid);
+            model.getModelMap().put("myProjectId", projid);
```

Comment:
move these two lines up before the if statement, they are repetitive. You can instantiate the model with an empty constructor and set the view name through a setter (or inject the model and return a string).